### PR TITLE
JAX time series classification

### PR DESCRIPTION
**Pull request by mtsokol**
This PR adds a tutorial with JAX time series classification task on a FordA dataset from the [UCR/UEA archive](https://www.cs.ucr.edu/%7Eeamonn/time_series_data_2018/). It follows one of the [Keras tutorials](https://keras.io/examples/timeseries/timeseries_classification_from_scratch/). It's still missing the narrative.

### CODE REVIEW

**Comment by mtsokol** [comment]
It's ready for a review!

**Review by jakevdp — APPROVED** [review]
Looks good – it's a bit sparse on the description of what the code is doing, but we can probably address that in a later pass.

**Comment by jakevdp** [comment]
Please resolve the conflicts, and then we can merge this. Thanks!

**Comment by mtsokol** [comment]
> Please resolve the conflicts, and then we can merge this. Thanks!

Rebased!

I also added a couple of descriptions before some of the code cells.

File: docs/JAX_time_series_classification.md [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+train_metrics_history = {
+    "train_loss": [],
+}
+
+eval_metrics_history = {
+    "test_loss": [],
+    "test_accuracy": [],
+}
+```
+
```
Comment (pavithraes) [inline docs/JAX_time_series_classification.md:283]:
```suggestion

We can now train the CNN model.
We'll evaluate the model’s performance on the test set after each epoch,
and print the metrics: total loss and accuracy.

```

File: docs/JAX_time_series_classification.md [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+        print(f"- Accuracy: {eval_metrics_history['test_accuracy'][-1]:0.4f}")
+```
+
+```{code-cell} ipython3
+%%time
+for epoch in range(num_epochs):
+    train_one_epoch(epoch)
+    evaluate_model(epoch)
+```
+
```
Comment (pavithraes) [inline docs/JAX_time_series_classification.md:326]:
```suggestion

Finally, let's visualize the loss and accuracy with Matplotlib.

```

File: docs/JAX_time_series_classification.md [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+axs[1].plot(eval_metrics_history["test_accuracy"])
+```
+
+Our model reached almost 90% accuracy on the test set after 300 epochs, but it's worth noting
+that the loss function isn't completely flat yet. We could continue until the curve flattens,
+but we also need to pay attention to validation accuracy so as to spot when the model starts
+overfitting.
+
+For model early stopping and selecting best model there's [Orbax](https://github.com/google/orbax)
+library which provides checkpointing and persistence utilities.
```
Comment (pavithraes) [inline docs/JAX_time_series_classification.md:346]:
```suggestion
For model early stopping and selecting best model, you can check out [Orbax](https://github.com/google/orbax),
a library which provides checkpointing and persistence utilities.
```

File: docs/JAX_time_series_classification.md [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+kernelspec:
+  display_name: jax-env
+  language: python
+  name: python3
+---
+
+# Time series classification with JAX
+
+In this tutorial, we're going to perform time series classification with a Convolutional Neural Network.
+We're going to use FordA dataset from the [UCR archive](https://www.cs.ucr.edu/%7Eeamonn/time_series_data_2018/).
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply docs/JAX_time_series_classification.md]: Done!

File: docs/JAX_time_series_classification.md [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+
+The problem we're facing is to assess if an engine is malfunctioning based on recorded noises it generates.
+Each sample is comprised of noise measurements across time, together with a "yes/no" label, so it's a binary classification problem.
+
+Although convolution models are mainly associated with image processing, they are useful also for time series data as they're able to extract temporal structures.
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply docs/JAX_time_series_classification.md]: Done!

File: docs/JAX_time_series_classification.md [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+## Tools overview
+
+Here's a list of key packages that belong to JAX AI stack:
+
+- [JAX](https://github.com/jax-ml/jax) will be used for array computations.
+- [Flax](https://github.com/google/flax) for constructing neural networks.
+- [Optax](https://github.com/google-deepmind/optax) for gradient processing and optimization.
+- [Grain](https://github.com/google/grain/) will be be used to define data sources.
+- [tqdm](https://tqdm.github.io/) for a progress bar to monitor the training progress.
+
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply docs/JAX_time_series_classification.md]: Done!

File: docs/JAX_time_series_classification.md [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+  name: python3
+---
+
+# Time series classification with JAX
+
+In this tutorial, we're going to perform time series classification with a Convolutional Neural Network.
+We're going to use FordA dataset from the [UCR archive](https://www.cs.ucr.edu/%7Eeamonn/time_series_data_2018/).
+
+The problem we're facing is to assess if an engine is malfunctioning based on recorded noises it generates.
+Each sample is comprised of noise measurements across time, together with a "yes/no" label, so it's a binary classification problem.
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply docs/JAX_time_series_classification.md]: Done!

File: docs/JAX_time_series_classification.md [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+import numpy as np
+import matplotlib.pyplot as plt
+import grain.python as grain
+import tqdm
+```
+
+## Dataset
+
+We load dataset files into NumPy arrays, add singleton dimention to take into
+the account convolution features, and change `-1` label to `0` value:
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply docs/JAX_time_series_classification.md]: Done!

File: docs/JAX_time_series_classification.md [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+
+    return (x_train, y_train), (x_test, y_test)
+```
+
+```{code-cell} ipython3
+(x_train, y_train), (x_test, y_test) = prepare_ucr_dataset()
+```
+
+```{code-cell} ipython3
+# Here are exemplary samples from each class
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply docs/JAX_time_series_classification.md]: Done!

File: docs/JAX_time_series_classification.md [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+from flax import nnx
+import optax
+
+import numpy as np
+import matplotlib.pyplot as plt
+import grain.python as grain
+import tqdm
+```
+
+## Dataset
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply docs/JAX_time_series_classification.md]: Done!

File: docs/JAX_time_series_classification.md [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+```{code-cell} ipython3
+# Here are exemplary samples from each class
+classes = np.unique(np.concatenate((y_train, y_test), axis=0))
+for c in classes:
+    c_x_train = x_train[y_train == c]
+    plt.plot(c_x_train[0], label="class " + str(c))
+plt.legend()
+plt.show()
+```
+
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply docs/JAX_time_series_classification.md:103]: Done!

File: docs/JAX_time_series_classification.md [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+    def __len__(self):
+        return len(self._x)
+```
+
+```{code-cell} ipython3
+train_source = DataSource(x_train, y_train)
+test_source = DataSource(x_test, y_test)
+```
+
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply docs/JAX_time_series_classification.md:133]: Done!

File: docs/JAX_time_series_classification.md [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+)
+
+test_sampler = grain.IndexSampler(
+    len(test_source),
+    shuffle=False,
+    seed=seed,
+    shard_options=grain.NoSharding(),  # No sharding since this is a single-device setup
+    num_epochs=1,                      # Iterate over the dataset for one epoch
+)
+
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply docs/JAX_time_series_classification.md:161]: Done!

File: docs/JAX_time_series_classification.md [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+plt.show()
+```
+
+For handling input data we're going to use Grain, a pure Python package developed
+for JAX and Flax models. Grain supports custom setups where data sources might come
+in different forms, but they all need to implement the `grain.RandomAccessDataSource`
+interface. See [PyGrain Data Sources](https://github.com/google/grain/blob/main/docs/data_sources.md)
+for more details.
+
+Our dataset is comprised of relatively small NumPy arrays so our DataSource is uncomplicated:
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply docs/JAX_time_series_classification.md]: Done!

File: docs/JAX_time_series_classification.md [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+    sampler=test_sampler,  # Sampler to determine how to access the data
+    worker_count=4,        # Number of child processes launched to parallelize the transformations among
+    worker_buffer_size=2,  # Count of output batches to produce in advance per worker
+    operations=[
+        grain.Batch(test_batch_size),
+    ]
+)
+```
+
+## Model
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply docs/JAX_time_series_classification.md]: Done!

File: docs/JAX_time_series_classification.md [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+
+Here we construct the model with three convolution and dense layers. We use ReLU activation
+function for middle layers and softmax in the final layer for binary classification output:
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply docs/JAX_time_series_classification.md]: Done!

File: docs/JAX_time_series_classification.md [hunk 17]
```diff
@@ -0,0 +1,346 @@
...
+for Deep Learning models:
+
+```{code-cell} ipython3
+num_epochs = 300
+learning_rate = 0.0005
+momentum = 0.9
+
+optimizer = nnx.Optimizer(model, optax.adam(learning_rate, momentum))
+```
+
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply docs/JAX_time_series_classification.md:253]: Done!

File: docs/JAX_time_series_classification.md [hunk 18]
```diff
@@ -0,0 +1,346 @@
...
+```{code-cell} ipython3
+def compute_losses_and_logits(model: nnx.Module, batch_tokens: jax.Array, labels: jax.Array):
+    logits = model(batch_tokens)
+
+    loss = optax.softmax_cross_entropy_with_integer_labels(
+        logits=logits, labels=labels
+    ).mean()
+    return loss, logits
+```
+
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply docs/JAX_time_series_classification.md:266]: Done!

File: docs/JAX_time_series_classification.md (hunk 1 above)
Comment (mtsokol) [reply docs/JAX_time_series_classification.md:321]: Done!

File: docs/JAX_time_series_classification.md (hunk 3 above)
Comment (mtsokol) [reply docs/JAX_time_series_classification.md]: Done!

**Comment by trallard** [comment]
@jakevdp this should be ready now 🚀

File: docs/JAX_time_series_classification.md [hunk 19]
```diff
@@ -0,0 +1,389 @@
...
+        x = jnp.mean(x, axis=(1,))  # global average pooling
+        x = self.dense_1(x)
+        x = jax.nn.softmax(x)
+        return x
+```
+
+```{code-cell} ipython3
+model = MyModel(rngs=nnx.Rngs(0))
+nnx.display(model)
```
Comment (jakevdp) [inline docs/JAX_time_series_classification.md:233]: This cell is producing a GPU warning – let's remove it from the notebook output to avoid distracting readers.

Comment (mtsokol) [reply docs/JAX_time_series_classification.md:233]: Done!

I squashed the PR, it's ready from my side!
